`object-shorthand` and `quote-props` have incompatibilities

```bash
$> eslint -v
v1.2.1
```

My `.eslintrc`:

```json
{
  "ecmaFeatures": {
    "objectLiteralShorthandProperties": true
  },
  "rules": {
    "quote-props": [2, "consistent"],
    "object-shorthand": [2, "always"]
  }
}
```

The code:

```javascript
var x = true;
({
  x,
  'needs-quote': true
});
```

Running `eslint` gives the following output:

```bash
index.js
  2:2  error  Inconsistently quoted property `needs-quote` found  quote-props
```

Changing the code to:

```javascript
var x = true;
({
  'x': x,
  'needs-quote': true
});
```

Will instead give the error:

```bash
index.js
  3:3  error  Expected property shorthand  object-shorthand
```

So there's a Catch-22 between these two rules. I could of course rename the `x` variable to something else, but I would prefer to be able to use the object property shorthand syntax even when I have `"consistent"` or `"consistent-as-needed"` for the `quote-props` rule.

What do you think? I could probably solve this myself and submit a PR but the question is what's actually the best solution for this? Or maybe you're of the opinion that this is by design and the rules should really be incompatible when using this combination of options?

Thank you for the report.

I think this is a bug of `quote-props`.
This rule should ignore computed/shorthand properties.

```js
/* eslint-env es6 */
/* eslint quote-props: [2, "always"] */
var x;
var obj = ({
    [x]: 0,
    x
});
```

Actual:

```
> eslint -v
v1.2.1
> eslint test.js --no-eslintrc

test.js
  5:5  error  Unquoted property `x` found  quote-props
  6:5  error  Unquoted property `x` found  quote-props

✖ 2 problems (2 errors, 0 warnings)
```

Expected:

no errors.

@BYK My intent is "This rule should ignore shorthand properties."

```
/* eslint-env es6 */
/* eslint quote-props: [2, "consistent"] */
var x;
var obj1 = ({
    "a-b": 0,
    x
});
var obj2 = ({
    x,
    "a-b": 0
});
```

```
> eslint test.js --no-eslintrc

test.js
  4:13  error  Inconsistently quoted property `x` found    quote-props
  8:13  error  Inconsistently quoted property `a-b` found  quote-props

✖ 2 problems (2 errors, 0 warnings)
```

I thought this is an unexpected behavior. Because `x` is an [IdentifierReference, not PropertyName](http://www.ecma-international.org/ecma-262/6.0/#sec-object-initializer).

--- a/lib/rules/quote-props.js
+++ b/lib/rules/quote-props.js
@@ -59,6 +59,10 @@ module.exports = function(context) {
             isKeywordToken,
             tokens;
 
+        if (node.method || node.computed || node.shorthand) {
+            return;
+        }
+
         if (key.type === "Literal" && typeof key.value === "string") {
             try {
                 tokens = espree.tokenize(key.value);
@@ -92,7 +96,7 @@ module.exports = function(context) {
     function checkOmittedQuotes(node) {
         var key = node.key;
 
-        if (!node.method && !(key.type === "Literal" && typeof key.value === "string")) {
+        if (!node.method && !node.computed && !node.shorthand && !(key.type === "Literal" && typeof key.value === "string")) {
             context.report(node, MESSAGE_UNQUOTED, {
                 property: key.name || key.value
             });
@@ -113,6 +117,10 @@ module.exports = function(context) {
         node.properties.forEach(function(property) {
             var key = property.key,
                 tokens;
+
+            if (property.method || property.computed || property.shorthand) {
+                return;
+            }
 
             if (key.type === "Literal" && typeof key.value === "string") {
 

--- a/tests/lib/rules/quote-props.js
+++ b/tests/lib/rules/quote-props.js
@@ -26,7 +26,13 @@ ruleTester.run("quote-props", rule, {
         "({ '@': 0 })",
 
         { code: "({ 'a': 0, b(){} })", ecmaFeatures: { objectLiteralShorthandMethods: true }},
+        { code: "({ [x]: 0 });", env: {es6: true}},
+        { code: "({ x });", env: {es6: true}},
         { code: "({ a: 0, b(){} })", options: ["as-needed"], ecmaFeatures: { objectLiteralShorthandMethods: true } },
+        { code: "({ a: 0, [x]: 1 })", options: ["as-needed"], env: {es6: true} },
+        { code: "({ a: 0, x })", options: ["as-needed"], env: {es6: true} },
+        { code: "({ '@': 0, [x]: 1 })", options: ["as-needed"], env: {es6: true} },
+        { code: "({ '@': 0, x })", options: ["as-needed"], env: {es6: true} },
         { code: "({ a: 0, b: 0 })", options: ["as-needed"] },
         { code: "({ a: 0, 0: 0 })", options: ["as-needed"] },
         { code: "({ a: 0, true: 0 })", options: ["as-needed"] },
@@ -42,10 +48,16 @@ ruleTester.run("quote-props", rule, {
         { code: "({ 'true': 0, 'b': 0 })", options: ["consistent"] },
         { code: "({ null: 0, a: 0 })", options: ["consistent"] },
         { code: "({ a: 0, b: 0 })", options: ["consistent"] },
+        { code: "({ 'a': 1, [x]: 0 });", options: ["consistent"], env: {es6: true}},
+        { code: "({ 'a': 1, x });", options: ["consistent"], env: {es6: true}},
         { code: "({ a: 0, b: 0 })", options: ["consistent-as-needed"] },
         { code: "({ a: 0, null: 0 })", options: ["consistent-as-needed"] },
         { code: "({ 'a': 0, '-b': 0 })", options: ["consistent-as-needed"] },
         { code: "({ '@': 0, 'B': 0 })", options: ["consistent-as-needed"] },
+        { code: "({ '@': 1, [x]: 0 });", env: {"es6": true}, options: ["consistent-as-needed"]},
+        { code: "({ '@': 1, x });", env: {"es6": true}, options: ["consistent-as-needed"]},
+        { code: "({ a: 1, [x]: 0 });", env: {"es6": true}, options: ["consistent-as-needed"]},
+        { code: "({ a: 1, x });", env: {"es6": true}, options: ["consistent-as-needed"]},
         { code: "({ a: 0, 'if': 0 })", options: ["as-needed", {keywords: true}] },
         { code: "({ a: 0, 'while': 0 })", options: ["as-needed", {keywords: true}] },
         { code: "({ a: 0, 'volatile': 0 })", options: ["as-needed", {keywords: true}] },
@@ -105,6 +117,20 @@ ruleTester.run("quote-props", rule, {
         }]
     }, {
         code: "({ 'a': 0, 'b': 0 })",
+        options: ["consistent-as-needed"],
+        errors: [{
+            message: "Properties shouldn't be quoted as all quotes are redundant.", type: "ObjectExpression"
+        }]
+    }, {
+        code: "({ 'a': 0, [x]: 0 })",
+        env: {"es6": true},
+        options: ["consistent-as-needed"],
+        errors: [{
+            message: "Properties shouldn't be quoted as all quotes are redundant.", type: "ObjectExpression"
+        }]
+    }, {
+        code: "({ 'a': 0, x })",
+        env: {"es6": true},
         options: ["consistent-as-needed"],
         errors: [{
             message: "Properties shouldn't be quoted as all quotes are redundant.", type: "ObjectExpression"

